interpreter/astwalker.go

```diff
@@ -0,0 +1,498 @@
+		// e.g, and(a, b, c) === a && b && c.
+		// Ensure the groups are appropriately laid-out in memory.
+		for i, argGroup := range argGroups {
+			evalCount += argGroupLens[i]
+			instructions = append(instructions, argGroup...)
+			if i != argCount-1 {
+				instructions = append(instructions,
+					NewJump(
+						argIds[i],
+						instructionCount-evalCount,
+						function == operators.LogicalOr))
+			}
+			evalCount += 1
+		}
+		return append(instructions, NewCall(node.Id, call.Function, argIds))
+
+	case operators.Conditional:
+		// Compute the conditional jump, with two jumps, one for false,
+		// and one for true
+		//
+		// Instruction layout:
+		// 0: condition
+		// 1: jump to <END> on undefined/error
+		// 2: jump to <ELSE> on false
+		// 3: <IF> expr
+		// 4: jump to <END>
+		// 5: <ELSE> expr
+		// 6: <END> ternary
+		conditionId, condition := argIds[0], argGroups[0]
+		trueId, trueVal := argIds[1], argGroups[1]
+		falseVal := argGroups[2]
+
+		// 0: condition
+		instructions = append(instructions, condition...)
+		// 1: jump to <END> on undefined/error
+		instructions = append(instructions,
+			NewJump(conditionId, len(trueVal)+len(falseVal)+3, unknownError))
+		// 2: jump to <ELSE> on false.
+		instructions = append(instructions,
+			NewJump(conditionId, len(trueVal)+2, false))
+		// 3: <IF> expr
+		instructions = append(instructions, trueVal...)
+		// 4: jump to <END>
+		instructions = append(instructions,
+			NewJump(trueId, len(falseVal), nil))
+		// 5: <ELSE> expr
+		instructions = append(instructions, falseVal...)
+		// 6: <END> ternary
+		return append(instructions, NewCall(node.Id, call.Function, argIds))
+
+	default:
+		for _, argGroup := range argGroups {
+			instructions = append(instructions, argGroup...)
+		}
+		return append(instructions, NewCall(node.Id, call.Function, argIds))
+	}
+}
+
+func (w *astWalker) walkList(node *expr.Expr) []Instruction {
+	listExpr := node.GetListExpr()
+	var elementIds []int64
+	var elementSteps []Instruction
+	for _, elem := range listExpr.GetElements() {
+		elementIds = append(elementIds, w.getId(elem))
+		elementSteps = append(elementSteps, w.walk(elem)...)
+	}
+	return append(elementSteps, NewList(node.Id, elementIds))
+}
+
+func (w *astWalker) walkStruct(node *expr.Expr) []Instruction {
+	structExpr := node.GetStructExpr()
+	keyValues := make(map[int64]int64)
+	fieldValues := make(map[string]int64)
+	var entrySteps []Instruction
+	for _, entry := range structExpr.GetEntries() {
+		valueId := w.getId(entry.GetValue())
+		switch entry.KeyKind.(type) {
+		case *expr.Expr_CreateStruct_Entry_FieldKey:
+			fieldValues[entry.GetFieldKey()] = valueId
+		case *expr.Expr_CreateStruct_Entry_MapKey:
+			keyValues[w.getId(entry.GetMapKey())] = valueId
+			entrySteps = append(entrySteps, w.walk(entry.GetMapKey())...)
+		}
+		entrySteps = append(entrySteps, w.walk(entry.GetValue())...)
+	}
+	if len(structExpr.MessageName) == 0 {
+		return append(entrySteps, NewMap(node.Id, keyValues))
+	} else {
```

Comment:
It would be more idiomatic to drop the `else`.

Reply:
Done